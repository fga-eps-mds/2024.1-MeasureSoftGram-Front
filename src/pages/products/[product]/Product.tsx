import React from 'react';
import Head from 'next/head';
import { useRouter } from 'next/router';
import Button from '@mui/material/Button';
import Typography from '@mui/material/Typography';

import { NextPageWithLayout } from '@pages/_app.next';
import getLayout from '@components/Layout';
import useRequireAuth from '@hooks/useRequireAuth';
import { useProductContext } from '@contexts/ProductProvider';

import { useQuery } from './hooks/useQuery';
import ProductContent from './components/ProductContent';

const Product: NextPageWithLayout = () => {
  useRequireAuth();
  const { repositoriesTsqmiHistory } = useQuery();
  const { currentProduct } = useProductContext();
  const router = useRouter();

  const handleViewRepositories = () => {
    router.push(`/products/${router.query.product}/repositories`);
  };

  const handleViewReleases = () => {
    router.push(`/products/${router.query.product}/releases`);
  };

  return (
    <>
      <Head>
        <title>{currentProduct?.name}</title>
      </Head>

      <ProductContent repositoriesTsqmiHistory={repositoriesTsqmiHistory} />

      <div style={{ margin: '20px', paddingLeft: '15px' }}>
        <div style={{ marginBottom: '20px' }}>
          <Typography variant="h5" gutterBottom>
            Repositórios
          </Typography>
          <Button variant="contained" color="primary" onClick={handleViewRepositories}>
            VER REPOSITÓRIOS
          </Button>
        </div>

        <div style={{ marginBottom: '20px' }}>
          <Typography variant="h5" gutterBottom>
            Releases
          </Typography>
          <Button variant="contained" color="primary" onClick={handleViewReleases}>
            VER RELEASES
          </Button>
        </div>
      </div>

    </>
  );
};

Product.getLayout = getLayout;

export default Product;
