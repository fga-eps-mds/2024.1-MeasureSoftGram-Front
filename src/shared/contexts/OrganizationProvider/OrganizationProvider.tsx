import React, { createContext, useState, useContext, ReactNode, useMemo, useEffect } from 'react';
import { Organization } from '@customTypes/organization';
import { organizationQuery } from '@services/organization';
import { toast } from 'react-toastify';
import { useAuth } from '@contexts/Auth';

interface Props {
  children: ReactNode;
}

interface IOrganizationContext {
  currentOrganization: Organization | null;
  currentOrganizations: Organization[];
  setCurrentOrganizations: (organizations: Organization[]) => void;
  organizationList: Organization[];
  isLoading: boolean;
  fetchOrganizations: (forceFetch?: boolean) => void;
}

const OrganizationContext = createContext<IOrganizationContext | undefined>(undefined);

export function OrganizationProvider({ children }: Props) {
  const { session } = useAuth();
  const [currentOrganizations, setCurrentOrganizations] = useState<Organization[]>([]);
  const [currentOrganization, setCurrentOrganization] = useState<Organization | null>(null);
  const [organizationList, setOrganizationList] = useState<Organization[]>([]);
  const [isLoading, setIsLoading] = useState(true);

  const fetchOrganizations = async (forceFetch?: boolean) => {
    if (!session && !forceFetch) return;
    setIsLoading(true);
    try {
      const result = await organizationQuery.getAllOrganization();
      if (result.type === 'success') {
        const organizations = result.value.map(org => ({
          id: org.id ?? '',
          name: org.name,
          description: org.description ?? '',
          url: org.url ?? '',
          products: org.products ?? [],
          key: org.key ?? ''
        }));
        setOrganizationList(organizations);
      } else {
        toast.error("Erro ao carregar organizações.");
      }
    } catch (error) {
      console.error("Failed to fetch organizations:", error);
      toast.error("Erro ao carregar organizações. Por favor, tente novamente.");
    } finally {
      setIsLoading(false);
    }
  };

  useEffect(() => {
    fetchOrganizations();
  }, []);

  useEffect(() => {
    if (organizationList.length > 0 && currentOrganizations.length === 0) {
      setCurrentOrganizations([organizationList[0]]);
    }
  }, [organizationList, currentOrganizations]);

  useEffect(() => {
    if (currentOrganizations.length > 0) {
      setCurrentOrganization(currentOrganizations[0]);
    } else {
      setCurrentOrganization(null);
    }
  }, [currentOrganizations]);

  const value = useMemo(() => {
    return {
      currentOrganization,
      currentOrganizations,
      setCurrentOrganizations,
      organizationList,
      isLoading,
      fetchOrganizations
    };
  }, [currentOrganization, currentOrganizations, organizationList, isLoading]);

  return <OrganizationContext.Provider value={value}>{children}</OrganizationContext.Provider>;
}

export function useOrganizationContext() {
  const context = useContext(OrganizationContext);

  if (context === undefined) {
    throw new Error('OrganizationContext must be used within a OrganizationProvider');
  }

  return context;
}
