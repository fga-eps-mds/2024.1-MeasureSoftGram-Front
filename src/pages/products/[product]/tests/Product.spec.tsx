import React from 'react';
import { render } from '@testing-library/react';

import { OrganizationProvider } from '@contexts/OrganizationProvider';
import Product from '../Product';

jest.mock('@contexts/ProductProvider', () => ({
  useProductContext: () => ({
    currentProduct: {
      name: `aoba`
    }
  })
}));

jest.mock('@contexts/RepositoryProvider', () => ({
  useRepositoryContext: () => ({})
}));

jest.mock('next/router', () => ({
  useRouter: () => ({
    query: { product: "1-5-MeasureSoftGram" },
    push: () => jest.fn(),
  })
}));

describe('Product', () => {
  describe('Snapshot', () => {
    it('Deve corresponder ao Snapshot', () => {
      const tree = render(
        <OrganizationProvider>
          <Product />
        </OrganizationProvider>
      );

      expect(tree).toMatchSnapshot();
    });
  });
});
