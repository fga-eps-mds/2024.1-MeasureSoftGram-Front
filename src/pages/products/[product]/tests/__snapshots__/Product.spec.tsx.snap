// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Snapshot Deve corresponder ao Snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
      >
        <div
          class="MuiBox-root css-j7qwjs"
        >
          <div
            class="MuiBox-root css-icv5vm"
          >
            <div
              class="MuiBox-root css-0"
            >
              <span
                class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1e1x4vv-MuiSkeleton-root"
                style="width: 400px; height: 50px;"
              />
              <span
                class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1e1x4vv-MuiSkeleton-root"
                style="width: 200px; height: 16px;"
              />
            </div>
          </div>
          <div
            class="MuiBox-root css-m76iwz"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1e1x4vv-MuiSkeleton-root"
              style="width: 150px; height: 45px;"
            />
          </div>
          <span
            class="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-pulse css-153n0d3-MuiSkeleton-root"
            style="height: 480px;"
          />
        </div>
      </div>
      <div
        style="margin: 20px; padding-left: 15px;"
      >
        <div
          style="margin-bottom: 20px;"
        >
          <h5
            class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
          >
            repo
          </h5>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            repo-button
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          style="margin-bottom: 20px;"
        >
          <h5
            class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
          >
            release
          </h5>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            release-button
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
    >
      <div
        class="MuiBox-root css-j7qwjs"
      >
        <div
          class="MuiBox-root css-icv5vm"
        >
          <div
            class="MuiBox-root css-0"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1e1x4vv-MuiSkeleton-root"
              style="width: 400px; height: 50px;"
            />
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1e1x4vv-MuiSkeleton-root"
              style="width: 200px; height: 16px;"
            />
          </div>
        </div>
        <div
          class="MuiBox-root css-m76iwz"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1e1x4vv-MuiSkeleton-root"
            style="width: 150px; height: 45px;"
          />
        </div>
        <span
          class="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-pulse css-153n0d3-MuiSkeleton-root"
          style="height: 480px;"
        />
      </div>
    </div>
    <div
      style="margin: 20px; padding-left: 15px;"
    >
      <div
        style="margin-bottom: 20px;"
      >
        <h5
          class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
        >
          repo
        </h5>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          repo-button
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        style="margin-bottom: 20px;"
      >
        <h5
          class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
        >
          release
        </h5>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          release-button
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
